src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -221,93 +221,202 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
+                    }
+
+                    if (offset < length)
+                    {
+                        lengthToExamine = length - offset;
+                        goto SequentialScan;
+                    }
+                }
+            }
+            else if (Vector.IsHardwareAccelerated)
+            {
+                if (offset < length)
                 {
-                    // Get the highest multiple of Vector<ushort>.Count that is within the search space.
-                    // That will be how many times we iterate in the loop below.
-                    // This is equivalent to: length = Vector<ushort>.Count * ((int)(pEndCh - pCh) / Vector<ushort>.Count)
-                    length = (int)((pEndCh - pCh) & ~(Vector<ushort>.Count - 1));
+                    lengthToExamine = GetCharVectorSpanLength(offset, length);
 
-                    // Get comparison Vector
-                    Vector<ushort> vComparison = new Vector<ushort>(value);
+                    Vector<ushort> values = new Vector<ushort>(value);
 
-                    while (length > 0)
+                    while (lengthToExamine > offset)
                     {
-                        // Using Unsafe.Read instead of ReadUnaligned since the search space is pinned and pCh is always vector aligned
-                        Debug.Assert(((int)pCh & (Unsafe.SizeOf<Vector<ushort>>() - 1)) == 0);
-                        Vector<ushort> vMatches = Vector.Equals(vComparison, Unsafe.Read<Vector<ushort>>(pCh));
-                        if (Vector<ushort>.Zero.Equals(vMatches))
+                        var matches = Vector.Equals(values, LoadVector(ref searchSpace, offset));
+                        if (Vector<ushort>.Zero.Equals(matches))
                         {
-                            pCh += Vector<ushort>.Count;
-                            length -= Vector<ushort>.Count;
+                            offset += Vector<ushort>.Count;
                             continue;
                         }
+
                         // Find offset of first match
-                        return (int)(pCh - pChars) + LocateFirstFoundChar(vMatches);
+                        return (int)(offset + LocateFirstFoundChar(matches));
                     }
 
-                    if (pCh < pEndCh)
+                    if (offset < length)
                     {
-                        length = (int)(pEndCh - pCh);
+                        lengthToExamine = length - offset;
                         goto SequentialScan;
                     }
                 }
-
-                return -1;
-            Found3:
-                pCh++;
-            Found2:
-                pCh++;
-            Found1:
-                pCh++;
-            Found:
-                return (int)(pCh - pChars);
             }
+            return -1;
+        Found3:
+            return (int)(offset + 3);
```

Comment:
Minor nit: you can save the x86 64-bit prefix on these instructions by writing this as `return (int)offset + 3;`. Since the domain of _offset_ is guaranteed to be within `Int32` you're not going to overflow.

Reply:
offset is `nint` so on x86 the cast doesn't do anything anyway? Shouldn't have a 64-bit prefix?